File: libs/go/combo.go
package stonecharts

import (
	"fmt"
	"math"
	"strconv"
	"strings"
)

const comboPad = 0.2

func renderComboSVG(spec *ChartSpec) string {
	return renderCartesian(spec, "Combo", "band", comboMarks, true)
}

func comboSeriesKind(s *Series) string {
	if s.Type == "line" {
		return "line"
	}
	return "column"
}

func comboYpix(f *cartesianFrame, s *Series) func(float64) float64 {
	if s.YAxis == 1 && f.secondaryAxis != nil {
		return f.ypix2
	}
	return f.ypix
}

func comboMarks(f *cartesianFrame, p *strings.Builder) {
	if f.n <= 0 {
		return
	}

	bandWidth := f.bandWidth()
	groupW := bandWidth * (1 - comboPad)
	stacked := f.stacking == "normal" || f.stacking == "percent"

	kCol := 0
	for si := range f.spec.Series {
		if comboSeriesKind(&f.spec.Series[si]) == "column" {
			kCol++
		}
	}
	kSlots := kCol
	if stacked || !f.spec.groupingOn() {
		kSlots = 1
	}
	if kSlots <= 0 {
		kSlots = 1
	}
	barW := groupW / float64(kSlots)

	kcMap := make(map[int]int)
	kc := 0
	for si := range f.spec.Series {
		if comboSeriesKind(&f.spec.Series[si]) == "column" {
			kcMap[si] = kc
			kc++
		}
	}

	totals := make([]float64, f.n)
	if stacked {
		for si := range f.spec.Series {
			if comboSeriesKind(&f.spec.Series[si]) != "column" {
				continue
			}
			for i, v := range f.spec.Series[si].Data {
				if i < f.n {
					totals[i] += v
				}
			}
		}
	}

	positive := make([]float64, f.n)
	negative := make([]float64, f.n)

	for si := range f.spec.Series {
		s := &f.spec.Series[si]
		if comboSeriesKind(s) == "line" {
			emitComboLineSeries(f, p, si, s)
		} else {
			emitComboColumnSeries(f, p, si, s, groupW, barW, kcMap, stacked, totals, positive, negative)
		}
	}
}

func emitComboColumnSeries(f *cartesianFrame, p *strings.Builder, si int, s *Series, groupW, barW float64, kcMap map[int]int, stacked bool, totals, positive, negative []float64) {
	st := f.styles[si]
	ypix := comboYpix(f, s)
	baseline := ypix(0.0)
	kc := kcMap[si]

	p.WriteString(fmt.Sprintf(`<g class="sc-series" data-series="%d">`, si))
	for i, raw := range s.Data {
		if i >= f.n {
			break
		}
		cxBand := f.xpix(float64(i))
		var left, y, h float64
		if stacked {
			left = cxBand - groupW/2
			value := raw
			if f.stacking == "percent" {
				total := totals[i]
				if total == 0 {
					value = 0
				} else {
					value = raw / total * 100.0
				}
			}
			var bottomV, topV float64
			if value >= 0 {
				bottomV = positive[i]
				topV = bottomV + value
				positive[i] = topV
			} else {
				bottomV = negative[i]
				topV = bottomV + value
				negative[i] = topV
			}
			y0 := ypix(bottomV)
			y1 := ypix(topV)
			y = math.Min(y0, y1)
			h = math.Abs(y0 - y1)
		} else {
			slot := 0
			if f.spec.groupingOn() {
				slot = kc
			}
			left = cxBand - groupW/2 + barW*float64(slot)
			yv := ypix(raw)
			y = math.Min(baseline, yv)
			h = math.Abs(baseline - yv)
		}
		xlabel := strconv.Itoa(i)
		if i < len(f.cats) {
			xlabel = f.cats[i]
		}
		cx := left + barW/2
		common := fmt.Sprintf(
			`class="sc-bar sc-point" data-series="%d" data-series-name="%s" data-x="%s" data-y="%s" data-color="%s" data-r="3.5" data-r-hover="6"`,
			si, esc(s.Name), esc(xlabel), esc(fmtNum(raw)), st.solid)
		p.WriteString(fmt.Sprintf(
			`<rect %s cx="%s" cy="%s" x="%s" y="%s" width="%s" height="%s" fill="%s"/>`,
			common, f1(cx), f1(y), f1(left), f1(y), f1(barW), f1(h), st.fill))
	}
	p.WriteString(`</g>`)
}

func emitComboLineSeries(f *cartesianFrame, p *strings.Builder, si int, s *Series) {
	st := f.styles[si]
	ypix := comboYpix(f, s)
	theme := f.theme
	color := st.solid
	pts := make([][2]float64, len(s.Data))
	for i, v := range s.Data {
		pts[i] = [2]float64{f.xpix(float64(i)), ypix(v)}
	}
	var d string
	if s.Curve == "monotone" {
		d = splineD(pts)
	} else {
		d = pathD(pts, s.Step)
	}
	lineDashAttr := ""
	if da := dashArray(s.DashStyle); da != "" {
		lineDashAttr = ` stroke-dasharray="` + da + `"`
	}
	p.WriteString(fmt.Sprintf(`<g class="sc-series" data-series="%d">`, si))
	if st.areaFill != "" && len(pts) > 0 {
		base := ypix(0.0)
		areaD := d + " L" + f1(pts[len(pts)-1][0]) + " " + f1(base) +
			" L" + f1(pts[0][0]) + " " + f1(base) + " Z"
		p.WriteString(fmt.Sprintf(
			`<path class="sc-series-area" data-series="%d" d="%s" fill="%s"%s stroke="none"/>`,
			si, areaD, st.areaFill, st.areaOp))
	}
	p.WriteString(fmt.Sprintf(
		`<path class="sc-series-line" data-series="%d" d="%s" fill="none" stroke="%s" stroke-width="%s" stroke-linejoin="round" stroke-linecap="round"%s/>`,
		si, d, st.stroke, fmtNum(s.lineWidth()), lineDashAttr))
	if s.markerEnabled() {
		radius := s.markerRadius()
		radiusHover := radius + 2.5
		symbol := s.markerSymbol()
		for i, pt := range pts {
			xlabel := strconv.Itoa(i)
			if i < len(f.cats) {
				xlabel = f.cats[i]
			}
			common := fmt.Sprintf(
				`class="sc-point" data-series="%d" data-series-name="%s" data-x="%s" data-y="%s" data-color="%s" data-r="%s" data-r-hover="%s"`,
				si, esc(s.Name), esc(xlabel), esc(fmtNum(s.Data[i])), color, fmtNum(radius), fmtNum(radiusHover))
			p.WriteString(markerSVG(symbol, pt[0], pt[1], radius, common, color, theme.MarkerHalo, 1.0))
		}
	}
	p.WriteString(`</g>`)
}

File: libs/python/stonecharts/charts/combo.py
"""Combo chart renderer: ChartSpec -> SVG string.

Composition layer dispatching per-series to column or line marks on a shared
plot area. Shared chrome from _cartesian.py; mark logic from column and line.
"""

from __future__ import annotations

from ..spec import Marker
from ..util import esc, fmt_num
from ._cartesian import CartesianFrame, dash_array, render_cartesian
from .line import _marker, _path_d, _spline_d

PAD = 0.2


def _kind(s) -> str:
    return s.type if s.type else "column"


def _ypix_for(fr: CartesianFrame, s):
    return fr.ypix2 if s.y_axis == 1 and fr.secondary_axis is not None else fr.ypix


def render_svg(spec) -> str:
    return render_cartesian(spec, "Combo", "band", _combo_marks)


def _combo_marks(fr: CartesianFrame, p: list) -> None:
    if fr.n <= 0:
        return

    band_width = fr.band_width()
    group_w = band_width * (1 - PAD)
    stacked = fr.stacking in ("normal", "percent")

    col_indices = [si for si, s in enumerate(fr.spec.series) if _kind(s) == "column"]
    k_col = len(col_indices)
    k_slots = 1 if stacked or not fr.spec.grouping else max(k_col, 1)
    bar_w = group_w / k_slots

    kc_map: dict[int, int] = {}
    kc = 0
    for si, s in enumerate(fr.spec.series):
        if _kind(s) == "column":
            kc_map[si] = kc
            kc += 1

    totals = [0.0] * fr.n
    if stacked:
        for si in col_indices:
            s = fr.spec.series[si]
            for i, v in enumerate(s.data):
                if i < fr.n:
                    totals[i] += v

    positive = [0.0] * fr.n
    negative = [0.0] * fr.n

    for si, s in enumerate(fr.spec.series):
        if _kind(s) == "line":
            _emit_line_series(fr, p, si, s)
        else:
            _emit_column_series(
                fr,
                p,
                si,
                s,
                group_w,
                bar_w,
                kc_map,
                stacked,
                totals,
                positive,
                negative,
            )


def _emit_column_series(fr, p, si, s, group_w, bar_w, kc_map, stacked, totals, positive, negative):
    st = fr.styles[si]
    ypix = _ypix_for(fr, s)
    baseline = ypix(0.0)
    kc = kc_map[si]

    p.append(f'<g class="sc-series" data-series="{si}">')
    for i, raw in enumerate(s.data):
        if i >= fr.n:
            break
        cx_band = fr.xpix(i)
        if stacked:
            left = cx_band - group_w / 2
            if fr.stacking == "percent":
                total = totals[i]
                value = 0.0 if total == 0.0 else raw / total * 100.0
            else:
                value = raw
            if value >= 0:
                bottom_v = positive[i]
                top_v = bottom_v + value
                positive[i] = top_v
            else:
                bottom_v = negative[i]
                top_v = bottom_v + value
                negative[i] = top_v
            y0 = ypix(bottom_v)
            y1 = ypix(top_v)
            y = min(y0, y1)
            h = abs(y0 - y1)
        else:
            left = cx_band - group_w / 2 + bar_w * (kc if fr.spec.grouping else 0)
            yv = ypix(raw)
            y = min(baseline, yv)
            h = abs(baseline - yv)
        xlabel = fr.cats[i] if i < len(fr.cats) else str(i)
        cx = left + bar_w / 2
        common = (
            f'class="sc-bar sc-point" data-series="{si}" data-series-name="{esc(s.name)}" '
            f'data-x="{esc(xlabel)}" data-y="{esc(fmt_num(raw))}" '
            f'data-color="{st.solid}" data-r="3.5" data-r-hover="6"'
        )
        p.append(
            f'<rect {common} cx="{cx:.1f}" cy="{y:.1f}" x="{left:.1f}" y="{y:.1f}" '
            f'width="{bar_w:.1f}" height="{h:.1f}" fill="{st.fill}"/>'
        )
    p.append("</g>")


def _emit_line_series(fr, p, si, s):
    st = fr.styles[si]
    ypix = _ypix_for(fr, s)
    theme = fr.theme
    pts = [(fr.xpix(i), ypix(v)) for i, v in enumerate(s.data)]
    d = _spline_d(pts) if s.curve == "monotone" else _path_d(pts, s.step)
    lw = s.line_width if s.line_width is not None else 2
    line_dash = dash_array(s.dash_style)
    line_dash_attr = f' stroke-dasharray="{line_dash}"' if line_dash else ""
    p.append(f'<g class="sc-series" data-series="{si}">')
    if st.area_fill is not None and pts:
        base = ypix(0.0)
        area_d = f"{d} L{pts[-1][0]:.1f} {base:.1f} L{pts[0][0]:.1f} {base:.1f} Z"
        p.append(
            f'<path class="sc-series-area" data-series="{si}" d="{area_d}" '
            f'fill="{st.area_fill}"{st.area_op} stroke="none"/>'
        )
    p.append(
        f'<path class="sc-series-line" data-series="{si}" d="{d}" fill="none" '
        f'stroke="{st.stroke}" stroke-width="{fmt_num(lw)}" stroke-linejoin="round" '
        f'stroke-linecap="round"{line_dash_attr}/>'
    )
    mk = s.marker or Marker()
    if mk.enabled:
        radius = mk.radius
        radius_hover = radius + 2.5
        for i, (x, y) in enumerate(pts):
            xlabel = fr.cats[i] if i < len(fr.cats) else str(i)
            common = (
                f'class="sc-point" data-series="{si}" data-series-name="{esc(s.name)}" '
                f'data-x="{esc(xlabel)}" data-y="{esc(fmt_num(s.data[i]))}" '
                f'data-color="{st.solid}" data-r="{fmt_num(radius)}" '
                f'data-r-hover="{fmt_num(radius_hover)}"'
            )
            p.append(_marker(mk.symbol, x, y, radius, common, st.solid, theme.marker_halo))
    p.append("</g>")
